pkg/controller/ads/dns.go

```diff
@@ -0,0 +1,239 @@
+/*
+ * Copyright The Kmesh Authors.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at:
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+package ads
+
+import (
+	"net"
+	"net/netip"
+	"slices"
+
+	clusterv3 "github.com/envoyproxy/go-control-plane/envoy/config/cluster/v3"
+	v3 "github.com/envoyproxy/go-control-plane/envoy/config/core/v3"
+	endpointv3 "github.com/envoyproxy/go-control-plane/envoy/config/endpoint/v3"
+	"google.golang.org/protobuf/types/known/wrapperspb"
+	"k8s.io/client-go/util/workqueue"
+
+	core_v2 "kmesh.net/kmesh/api/v2/core"
+	"kmesh.net/kmesh/pkg/dns"
+)
+
+// adsDnsResolver is DNS resolver of Kernel Native
+type AdsDnsResolver struct {
+	Clusters        chan []*clusterv3.Cluster
+	adsCache        *AdsCache
+	dnsResolver     *dns.DNSResolver
+	dnsRefreshQueue workqueue.TypedDelayingInterface[any]
+}
+
+func NewAdsDnsResolver(adsCache *AdsCache) (*AdsDnsResolver, error) {
```

Comment:
It is under ads pkg, so we can remove the redundant

```suggestion
func NewDnsResolver(adsCache *AdsCache) (*DnsResolver, error) {
```